ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/OrphanArbitrary.hs

```diff
@@ -0,0 +1,50 @@
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Test.Ouroboros.Storage.LedgerDB.OrphanArbitrary where
+
+import           Test.QuickCheck
+import           Data.Time.Clock (DiffTime, secondsToDiffTime, diffTimeToPicoseconds)
+import           Ouroboros.Consensus.Config.SecurityParam (SecurityParam(..))
+import           Ouroboros.Consensus.Storage.LedgerDB.DiskPolicy (SnapshotInterval(..))
+
+{-------------------------------------------------------------------------------
+  Orphan Arbitrary instances
+-------------------------------------------------------------------------------}
+instance Arbitrary SecurityParam where
+  arbitrary = SecurityParam <$> choose (0, 6)
+  shrink (SecurityParam k) = SecurityParam <$> shrink k
+
+-------------------------------------------------------------------------------}
+
+instance Arbitrary SnapshotInterval where
+  arbitrary =
+    frequency [ (9,  requestedSnapshotInterval)
+              , (1, defaultSnapshotInterval)
+              ]
+  shrink DefaultSnapshotInterval = []
+  shrink (RequestedSnapshotInterval v) = RequestedSnapshotInterval <$> shrink v
+
+requestedSnapshotInterval :: Gen SnapshotInterval
+requestedSnapshotInterval =
+  RequestedSnapshotInterval <$> arbitrary
+
+defaultSnapshotInterval :: Gen SnapshotInterval
+defaultSnapshotInterval = pure DefaultSnapshotInterval
+
+newtype AlwaysDefaultSnapshotInterval = AlwaysDefaultSnapshotInterval SnapshotInterval
```

Comment:
This is what I had in mind. What do you think? An attractive advantage is that it avoids the need for `-fno-warn-incomplete-patterns`.

```
-- | The isolated 'DefaultSnapshotInterval' constructor
data TheDefaultSnapshotInterval = TheDefaultSnapshotInterval

-- | An isolated 'RequestedSnapshotInterval' constructor
newtype ARequestedSnapshotInterval = ARequestedSnapshotInterval DiffTime

fromTheDefaultSnapshotInterval :: TheDefaultSnapshotInterval -> RequestedSnapshotInterval
fromTheDefaultSnapshotInterval TheDefaultSnapshotInterval = DefaultSnapshotInterval

fromARequestedSnapshotInterval :: ARequestedSnapshotInterval -> RequestedSnapshotInterval
fromARequestedSnapshotInterval (ARequestedSnapshotInterval d) = RequestedSnapshotInterval d
```

And then the reuse in the `Arbitrary` instances is in the other direction.

```
instance Arbitrary TheDefaultSnapshotInterval where
  arbitrary = pure TheDefaultSnapshotInterval

instance Arbitrary ARequestedSnapshotInterval where
  arbitrary = ARequestedSnapshotInterval . fromIntegral <$> choose (1, 3600 :: Int)   -- seconds
  -- TODO shrink (see other comment thread about `Pico`)
  
instance Arbitrary SnapshotInterval where
  arbitrary =
    frequency [ (9, fromARequestedSnapshotInterval <$> arbitrary)
              , (1, fromTheDefaultSnapshotInterval <$> arbitrary)
              ]
  -- TODO shrink
```

Reply:
After giving it some thought (and some coding, I have a commit ready but not pushed yet) I must say I dislike this approach as compared to what I wrote. For multiple reasons:

1. A general (arbitrary) instance is written using definitions of the "problem-specific" instances.

It seems more natural that it should be the other way around. 
In my code an arbitrary instance was implemented using two helper functions `requestedSnapshotInterval` and `defaultSnapshotInterval` and only then the two more "problem-specific" instances were defined using one function each.

2. This approach does not remove the need for `-fno-warn-incomplete-patterns`

The only reason that `-fno-warn-incomplete-patterns` was introduce was in a single test where not only we required a specific data constructor (`RequestedSnapshotInterval`) but we were also interested in the *value* of that requested interval.

```
(AlwaysRequestedSnapshotInterval snapshotInterval@(RequestedSnapshotInterval interval))
-----------------------------------------------------------------------  HERE  ---- ^
```

So after applying your idea, the tests looks following:

```
prop_shouldSnapshot_case2 :: DiffTime -> Word64 -> SecurityParam -> ARequestedSnapshotInterval -> Property
prop_shouldSnapshot_case2 timeSinceLast blocksSinceLast securityParam aRequestedSnapshotInterval = do
  -- given
  let
    snapshotInterval = fromARequestedSnapshotInterval aRequestedSnapshotInterval
    interval = undefined --  no what? :) pattern matching? failing? no clue really
    diskPolicy = defaultDiskPolicy securityParam snapshotInterval
  -- when
    shouldSnapshot = (onDiskShouldTakeSnapshot diskPolicy) (Just timeSinceLast) blocksSinceLast
  -- then
  shouldSnapshot === (timeSinceLast >= interval) -- used here
  ```

Please note that in the equivalent test - where we requested the second data constructor (`DefaultSnapthotInterval`) but were not interested in the values of it (because those do not exists) - we would not need the `-fno-warn-incomplete-patterns` at all. This is ONLY the problem when we want to get access to the values of the generated data.

3. This encode seems more boiler-plate with the additional helper functions that are needed instead of doing plain old pattern matching.

Thoughts?

Reply:
2. For the `undefined`: `let ARequestedSnapshotInterval interval = aRequestedSnapshotInterval`

We can discuss 1 and 3 on the call.